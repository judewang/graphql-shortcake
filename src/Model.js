/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
