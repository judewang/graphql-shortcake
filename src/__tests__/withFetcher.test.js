import _ from 'lodash';
import { client } from 'knex';
import database from './database';
import Model from '..';

class Fetcher extends Model {
  static database = database;

  static tableName = 'fetcher';

  static hasOperator = true;

  static columns = {
    name: { type: String },
    data: { type: Object },
  }
}

describe('Fetcher', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('fetcher');

    await database.schema.createTable('fetcher', (table) => {
      table.increments();
      table.string('name');
      table.jsonb('data');
      table.timestamps();
      table.integer('created_by');
      table.integer('updated_by');
      table.timestamp('deleted_at');
      table.integer('deleted_by');
    });

    await new Fetcher().insert({ name: 'pineapple shortcake' });
    await new Fetcher().insert({ name: 'banana' });
  });

  describe('fetch', () => {
    it('fetch one', async () => {
      const model = new Fetcher({
        name: 'pineapple shortcake',
        data: { xyz: 1 },
      });
      await model.fetch();
      expect(model.values).toMatchSnapshot();
      expect(client).toMatchSnapshot();

      await expect((new Fetcher({ name: 'shortcake' })).fetch())
        .resolves.toEqual(null);
      await expect((new Fetcher({ name: 'shortcake' })).fetch(Error))
        .rejects.toEqual(new Error());
    });

    it('with where', async () => {
      const model = new Fetcher();
      model.where({ name: 'banana' });
      model.orderByAsc('createdAt');
      model.whereNull('data');
      await model.fetch();
      expect(model.values).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });
  });

  describe('fetchAll', () => {
    it('successfully fetch', async () => {
      const fetcher = new Fetcher({ name: 'banana' });

      const results = await fetcher.fetchAll();
      expect(results.totalCount).toBe(1);
      expect(results.offset).toBe(null);
      expect(results.limit).toBe(null);
      expect(_.map(results, 'name')).toEqual(['banana']);
      expect(client).toMatchSnapshot();
    });

    it('when not found', async () => {
      const fetcher = new Fetcher({ name: 'apple' });

      await expect(fetcher.fetchAll())
        .resolves.toEqual(Object.assign([], { totalCount: 0, offset: null, limit: null }));
      await expect(fetcher.fetchAll(Error))
        .rejects.toEqual(new Error());
    });
  });

  describe('fetchPage', () => {
    it('offset and first', async () => {
      const fetcher = new Fetcher();
      const results = await fetcher.fetchPage({ offset: 1, first: 2 });
      expect(results).toEqual(expect.objectContaining({ totalCount: 2, offset: 1, limit: 2 }));
      expect(_.map(results, 'name')).toEqual(['banana']);
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('not set', async () => {
      const model = new Fetcher();
      const results = await model.fetchPage();
      expect(results.totalCount).toBe(2);
      expect(results).toEqual(expect.objectContaining({ totalCount: 2, offset: 0, limit: 1000 }));
      expect(_.map(results, 'name')).toEqual(['pineapple shortcake', 'banana']);
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });
  });

  describe('fetchCount', () => {
    it('successfully fetch', async () => {
      const fetcher = new Fetcher({ name: 'pineapple shortcake' });
      await expect(fetcher.fetchCount()).resolves.toBe(1);
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('not found', async () => {
      const fetcher = new Fetcher({ name: 'cherry' });
      await expect(fetcher.fetchCount(Error))
        .rejects.toEqual(new Error());
    });
  });
});
