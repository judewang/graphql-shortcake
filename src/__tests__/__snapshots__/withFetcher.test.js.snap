// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetcher fetch fetch one 1`] = `
Object {
  "createdAt": null,
  "createdBy": null,
  "data": null,
  "deletedAt": null,
  "deletedBy": null,
  "id": 1,
  "name": "pineapple shortcake",
  "updatedAt": null,
  "updatedBy": null,
}
`;

exports[`Fetcher fetch fetch one 2`] = `
Array [
  "select * from fetcher where deleted_at is null and name = pineapple shortcake limit 1",
]
`;

exports[`Fetcher fetch with where 1`] = `Object {}`;

exports[`Fetcher fetch with where 2`] = `
Array [
  "select * from fetcher where name = shortcake and data IS NULL and deleted_at is null order by created_at asc limit 1",
]
`;

exports[`Fetcher fetchAll successfully fetch 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from fetcher where deleted_at is null and name = shortcake",
]
`;

exports[`Fetcher fetchCount successfully fetch 1`] = `
Array [
  "select count(*) from fetcher where deleted_at is null and name = pineapple shortcake",
]
`;

exports[`Fetcher fetchPage not set 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from fetcher where deleted_at is null limit 1000",
]
`;

exports[`Fetcher fetchPage offset and first 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from fetcher where deleted_at is null limit 2 offset 1",
]
`;

exports[`Fetcher saveIfNotExists when existed 1`] = `
Array [
  "select * from fetcher where deleted_at is null and name = pineapple shortcake limit 1",
]
`;

exports[`Fetcher saveIfNotExists when not existed 1`] = `
Array [
  "select * from fetcher where deleted_at is null and name = shortcake limit 1",
  "insert into fetcher (created_at, created_by, name, updated_at, updated_by) values (DATE, 10, shortcake, DATE, 10) returning *",
]
`;
